bindings/pyroot_experimental/PyROOT/src/DeclareCallablePyz.cxx

```diff
@@ -0,0 +1,663 @@
+      auto pytype = typemap.find(returnTypeStr);
+      if (pytype != typemap.end()) {
+         if (pytype->second == "f") {
+            code << "   auto result = PyFloat_AsDouble(pyresult);\n";
+         } else if (pytype->second == "i" || pytype->second == "l") {
+            code << "   auto result = PyLong_AsLong(pyresult);\n";
+         } else if (pytype->second == "I" || pytype->second == "k") {
+            code << "   auto result = PyLong_AsUnsignedLong(pyresult);\n";
+         }
+      } else {
+         code << "   if (!TPython::CPPInstance_Check(pyresult)) {\n"
+              << "      throw std::runtime_error(\"Failed to convert return value of Python callable to C++ object: Python object is not created by cppyy (CPPInstance).\");\n"
+              << "      \n"
+              << "   }\n";
+         code << "   auto result = *reinterpret_cast<" << returnTypeStr << "*>(TPython::CPPInstance_AsVoidPtr(pyresult));\n";
+      }
+
+      code << "   Py_DECREF(pyresult);\n\n"
+           << "   return result;\n";
+   }
+   code << "}\n}";
+
+   // Attach C++ wrapper code to callable
+   auto code_str = code.str();
+   auto code_cstr = code_str.c_str();
+   auto pycode = CPyCppyy_PyUnicode_FromString(code_cstr);
+   PyObject_SetAttrString(pyfunc, "__cpp_wrapper__", pycode);
+   Py_DECREF(pycode);
+
+   // Jit C++ wrapper
+   auto err = gInterpreter->Declare("#include \"Python.h\"");
+   if (!err) {
+      PyErr_SetString(PyExc_RuntimeError, "Failed to compile C++ wrapper: Failed to include Python.h.");
+      return NULL;
+   }
+
+   err = gInterpreter->Declare("#include \"TPython.h\"");
+   if (!err) {
+      PyErr_SetString(PyExc_RuntimeError, "Failed to compile C++ wrapper: Failed to include TPython.h.");
+      return NULL;
+   }
+
+   err = gInterpreter->Declare(code_cstr);
+   if (!err) {
+      PyErr_SetString(PyExc_RuntimeError,
+              ("Failed to compile C++ wrapper: Compilation error from following wrapper code.\n" + code.str()).c_str());
+      return NULL;
+   }
+
+   // Pass through Python callable
+   return pyfunc;
+}
+
+
+// Call method of class used as decorator to create C++ wrapper using numba
+// The call method creates the C++ wrapper class for the Python callable and
+// passes through the actual callable.
+PyObject* NumbaCallableImpl_call(PyObject * /*self*/, PyObject *args)
+{
+   // Parse arguments with Python callable and get attributes with types
+   if (!PyTuple_Check(args)) {
+      PyErr_SetString(PyExc_RuntimeError, "Failed to parse arguments: Invalid tuple.");
+      return NULL;
+   }
+
+   if (!(PyTuple_Size(args) == 2)) {
+      PyErr_SetString(PyExc_RuntimeError, "Failed to parse arguments: Expect exactly one argument (Python callable).");
+      return NULL;
+   }
+
+   if (!(PyTuple_Size(args) == 2)) {
+      PyErr_SetString(PyExc_RuntimeError, "Failed to parse arguments: Expect exactly one argument (Python callable).");
+      return NULL;
+   }
+
+   auto instance = PyTuple_GetItem(args, 0);
```

Comment:
This code here and below has quite some repetition with respect to `GenericCallableImpl_call`. If we just have one decorator, we can eliminate all this repetition.